src/hotspot/cpu/riscv/macroAssembler_riscv.cpp

```diff
@@ -3662,6 +3662,149 @@ void MacroAssembler::mul_add(Register out, Register in, Register offset,
+ *
+ * uint32_t crc;
+ * val = crc_table[(val ^ crc) & 0xFF];
+ * crc = val ^ (crc >> 8);
+ *
+ */
+void MacroAssembler::update_byte_crc32(Register crc, Register val, Register table) {
+  assert_different_registers(crc, val, table);
+
+  xorr(val, val, crc);
+  andi(val, val, bits8);
+  slli(val, val, 2);
+  add(val, table, val);
+  Assembler::lwu(val, val, 0);
+  srli(crc, crc, 8);
+  xorr(crc, val, crc);
+}
+
+/**
+ * Emits code to update CRC-32 with a 32-bit value according to tables 0 to 3
+ *
+ * @param [in,out]crc   Register containing the crc.
+ * @param [in]v         Register containing the 32-bit to fold into the CRC.
+ * @param [in]table0    Register containing table 0 of crc constants.
+ * @param [in]table1    Register containing table 1 of crc constants.
+ * @param [in]table2    Register containing table 2 of crc constants.
+ * @param [in]table3    Register containing table 3 of crc constants.
+ *
+ * uint32_t crc;
+ *   v = crc ^ v
+ *   crc = table3[v&0xff]^table2[(v>>8)&0xff]^table1[(v>>16)&0xff]^table0[v>>24]
+ *
+ */
+void MacroAssembler::update_word_crc32(Register crc, Register v, Register tmp1, Register tmp2,
+        Register table0, Register table1, Register table2, Register table3, bool upper) {
+  assert_different_registers(crc, v, tmp1, tmp2, table0, table1, table2, table3);
+
+  if (upper)
+    srli(v, v, 32);
+  xorr(v, crc, v);
+
+  andi(tmp1, v, bits8);
+  shadd(tmp1, tmp1, table3, tmp2, 2);
+  Assembler::lwu(crc, tmp1, 0);
+
+  srli(tmp1, v, 6);
+  andi(tmp1, tmp1, (bits8 << 2));
+  add(tmp1, tmp1, table2);
+  Assembler::lwu(tmp2, tmp1, 0);
+
+  srli(tmp1, v, 14);
+  xorr(crc, crc, tmp2);
+
+  andi(tmp1, tmp1, (bits8 << 2));
+  add(tmp1, tmp1, table1);
+  Assembler::lwu(tmp2, tmp1, 0);
+
+  srli(tmp1, v, 22);
+  xorr(crc, crc, tmp2);
+
+  andi(tmp1, tmp1, (bits8 << 2));
+  add(tmp1, tmp1, table0);
+  Assembler::lwu(tmp2, tmp1, 0);
+  xorr(crc, crc, tmp2);
+}
+
+/**
+ * @param crc   register containing existing CRC (32-bit)
+ * @param buf   register pointing to input byte buffer (byte*)
+ * @param len   register containing number of bytes
+ * @param table register that will contain address of CRC table
+ * @param tmp   scratch register
+ */
+void MacroAssembler::kernel_crc32(Register crc, Register buf, Register len,
+        Register table0, Register table1, Register table2, Register table3,
+        Register tmp, Register tmp2, Register tmp3, Register tmp4, Register tmp5) {
+  assert_different_registers(crc, buf, table0, table1, table2, table3, tmp, tmp2, tmp3, tmp4, tmp5);
+  Label L_by16, L_by16_loop, L_by16_loop_entry, L_by4, L_by4_loop, L_by1, L_by1_loop, L_exit;
+
+  mv(tmp5, bits32);
+  andn(crc, tmp5, crc);
+
+  const ExternalAddress table_addr = StubRoutines::crc_table_addr();
+  la(table0, table_addr);
+  add(table1, table0, 1*256*sizeof(juint), tmp);
+  add(table2, table0, 2*256*sizeof(juint), tmp);
+  add(table3, table0, 3*256*sizeof(juint), tmp);
```

Comment:
With `add(table3, table2, 256*(junit))`, it might save one instruction.

Reply:
Good catch, thanks! Fixed.